std/os/wasi.zig

```diff
@@ -0,0 +1,105 @@
+use @import("wasi/core.zig");
+
+pub const STDIN_FILENO = 0;
+pub const STDOUT_FILENO = 1;
+pub const STDERR_FILENO = 2;
+
+pub const ESUCCESS = 0;
+pub const E2BIG = 1;
+pub const EACCES = 2;
+pub const EADDRINUSE = 3;
+pub const EADDRNOTAVAIL = 4;
+pub const EAFNOSUPPORT = 5;
+pub const EAGAIN = 6;
+pub const EALREADY = 7;
+pub const EBADF = 8;
+pub const EBADMSG = 9;
+pub const EBUSY = 10;
+pub const ECANCELED = 11;
+pub const ECHILD = 12;
+pub const ECONNABORTED = 13;
+pub const ECONNREFUSED = 14;
+pub const ECONNRESET = 15;
+pub const EDEADLK = 16;
+pub const EDESTADDRREQ = 17;
+pub const EDOM = 18;
+pub const EDQUOT = 19;
+pub const EEXIST = 20;
+pub const EFAULT = 21;
+pub const EFBIG = 22;
+pub const EHOSTUNREACH = 23;
+pub const EIDRM = 24;
+pub const EILSEQ = 25;
+pub const EINPROGRESS = 26;
+pub const EINTR = 27;
+pub const EINVAL = 28;
+pub const EIO = 29;
+pub const EISCONN = 30;
+pub const EISDIR = 31;
+pub const ELOOP = 32;
+pub const EMFILE = 33;
+pub const EMLINK = 34;
+pub const EMSGSIZE = 35;
+pub const EMULTIHOP = 36;
+pub const ENAMETOOLONG = 37;
+pub const ENETDOWN = 38;
+pub const ENETRESET = 39;
+pub const ENETUNREACH = 40;
+pub const ENFILE = 41;
+pub const ENOBUFS = 42;
+pub const ENODEV = 43;
+pub const ENOENT = 44;
+pub const ENOEXEC = 45;
+pub const ENOLCK = 46;
+pub const ENOLINK = 47;
+pub const ENOMEM = 48;
+pub const ENOMSG = 49;
+pub const ENOPROTOOPT = 50;
+pub const ENOSPC = 51;
+pub const ENOSYS = 52;
+pub const ENOTCONN = 53;
+pub const ENOTDIR = 54;
+pub const ENOTEMPTY = 55;
+pub const ENOTRECOVERABLE = 56;
+pub const ENOTSOCK = 57;
+pub const ENOTSUP = 58;
+pub const ENOTTY = 59;
+pub const ENXIO = 60;
+pub const EOVERFLOW = 61;
+pub const EOWNERDEAD = 62;
+pub const EPERM = 63;
+pub const EPIPE = 64;
+pub const EPROTO = 65;
+pub const EPROTONOSUPPORT = 66;
+pub const EPROTOTYPE = 67;
+pub const ERANGE = 68;
+pub const EROFS = 69;
+pub const ESPIPE = 70;
+pub const ESRCH = 71;
+pub const ESTALE = 72;
+pub const ETIMEDOUT = 73;
+pub const ETXTBSY = 74;
+pub const EXDEV = 75;
+pub const ENOTCAPABLE = 76;
+
+// TODO: figure out what's going on here
```

Comment:
POSIX libc uses a thread-local variable `errno` for error codes. It's inefficient, confusing, and unnecessary. On Linux, syscalls use the range of the return code `-1, -2, -3, ... -4095` for error codes. Linux libc has this logic and then assigns it to `errno`. In Zig when not using libc we just look at the return code directly using this helper function.

The next step is looking at WASI specification to find out how errors are communicated in return codes. Based on https://github.com/CraneStation/wasi-sysroot/blob/wasi/libc-bottom-half/headers/public/wasi/core.h it looks to me like the functions directly return the error code in the return value, and use output pointer parameters for other return values.

So what you will actually want here is a way to put the errno back into the return code if you want to match Zig's POSIX API, which is what this file is - as far as WASI is concerned, this file is adapting the WASI API to Zig's flavor of the POSIX API. You can see an example of this with std/os/darwin.zig in the errnoWrap function.

Throughout the Zig standard library, I foresee the code having WASI-specific code and using the `wasi/core.zig` API directly. However it may be nice for users of the standard library to write code using the POSIX API and have it also work for WASI. Does that make sense?

Reply:
Yes, the functions return error codes. 

So that means `posix` under WASI will have a "global" error code somewhere, while majority of std will use core wasi API?

Reply:
To be clear, Zig's POSIX API layer does not match actual POSIX exactly. It follow's the Linux syscall convention of reserving a range of the return code integer for error codes. This part:

> So what you will actually want here is a way to put the errno back into the return code if you want to match Zig's POSIX API, which is what this file is - as far as WASI is concerned, this file is adapting the WASI API to Zig's flavor of the POSIX API. **You can see an example of this with std/os/darwin.zig in the errnoWrap function.**

For an initial implementation of WASI in the standard library, I don't even think we need to do the POSIX API layer. We could just open an issue for that and do it later. Most places in the standard library can use WASI core directly.